### PR TITLE
Fix websocket when reading small amount of data

### LINKED ISSUE #21989: websocket: recv_msg always returns full message length on last call
`websocket_recv_msg()` always returns full message length on last call:
https://github.com/zephyrproject-rtos/zephyr/blob/ba98e72db2f0c921ebb6a6d957a9c3609daec354/subsys/net/lib/websocket/websocket.c#L821

This is probably meant to limit the length to the actual message length when the buffer contains additional trailing data (of another message) and is fine when messages are entirely read at once.

However, using the MQTT client with `transport.type = MQTT_TRANSPORT_SECURE_WEBSOCKET` for example, when reading publish payload data in chunks as it is done in
https://github.com/zephyrproject-rtos/zephyr/blob/ba98e72db2f0c921ebb6a6d957a9c3609daec354/samples/net/cloud/google_iot_mqtt/src/protocol.c#L148-L151
then the `remaining_payload` variable will underflow in this line:
https://github.com/zephyrproject-rtos/zephyr/blob/ba98e72db2f0c921ebb6a6d957a9c3609daec354/subsys/net/lib/mqtt/mqtt.c#L661

Limiting `recv_len` by `recv_len = MIN(recv_len, ctx->message_len);` works for me.

### PR DESCRIPTION
**Pull request by jukkar**
Also add unit test for that.

Fixes #21989

### CODE REVIEW

File: subsys/net/lib/websocket/websocket.c [hunk 1]
```diff
@@ -443,12 +443,25 @@ static int websocket_ioctl_vmeth(void *obj, unsigned int request, va_list args)
...
+
+	/* Special handling when receiving data. As we cannot guarantee that
+	 * read data is received in four byte blocks, we need to pick proper
+	 * byte from the shift mask. When sending data, the mask_shift is set
+	 * to -1.
+	 */
+	if (mask_shift < 0) {
+		mask_shift_ptr = &zero;
+	} else {
+		mask_shift_ptr = &mask_shift;
```
Comment (pfalcon) [inline subsys/net/lib/websocket/websocket.c]: How var of type `int*` can be assigned with address of `s8_t`?

File: subsys/net/lib/websocket/websocket.c [hunk 2]
```diff
@@ -443,12 +443,25 @@ static int websocket_ioctl_vmeth(void *obj, unsigned int request, va_list args)
...
-				   u32_t masking_value)
+				   u32_t masking_value, s8_t mask_shift)
 {
-	int i;
+	int *mask_shift_ptr;
+	int i, zero = 0;
+
+	/* Special handling when receiving data. As we cannot guarantee that
+	 * read data is received in four byte blocks, we need to pick proper
+	 * byte from the shift mask. When sending data, the mask_shift is set
```
Comment (pfalcon) [inline subsys/net/lib/websocket/websocket.c]: > When sending data, the mask_shift is set to -1

Why do we call this routine at all "when sending data"?

File: subsys/net/lib/websocket/websocket.c [hunk 3]
```diff
@@ -443,12 +443,25 @@ static int websocket_ioctl_vmeth(void *obj, unsigned int request, va_list args)
 }
 
 static void websocket_mask_payload(u8_t *payload, size_t payload_len,
-				   u32_t masking_value)
+				   u32_t masking_value, s8_t mask_shift)
```
Comment (pfalcon) [inline subsys/net/lib/websocket/websocket.c]: Are you sure this argument should be of s8_t type, instead of int? Some compilers may produce less efficient code in thus case (effectively, adding sign extension cast from s8_t to full register size where all operations are handled then).

**Review by pfalcon — COMMENTED** [review]
Some type safety issues are evident.

File: subsys/net/lib/websocket/websocket.c (hunk 2 above)
Comment (pfalcon) [reply subsys/net/lib/websocket/websocket.c]: Ah, nevermind, it's the structure of the operations here got me confused here. I just remember that masking of data happens in one direction, but not another (e.g. server->client, but not vice-versa, I don't remember which exactly). So, the question then is why we need this special -1 value, why don't just pass 0 as an arg.

File: subsys/net/lib/websocket/websocket.c [hunk 4]
```diff
@@ -643,10 +656,24 @@ int websocket_recv_msg(int ws_sock, u8_t *buf, size_t buf_len,
 		       u32_t *message_type, u64_t *remaining, s32_t timeout)
 {
 	struct websocket_context *ctx;
-	size_t header_len, pos_to_write = 0;
+	size_t header_len = 0, pos_to_write = 0;
 	int recv_len = 0;
 	int ret;
 
+#if defined(CONFIG_NET_TEST)
+	/* Websocket unit test does not use socket layer but feeds
```
Comment (pfalcon) [inline subsys/net/lib/websocket/websocket.c]: I would image we'd end up with such "test handler" in the code eventually, e.g. for fault injection. But is it warranted here now? What tests you can't perform via socket layer which required such contraptions? In the end, users will user this API via socket layer, and that's what we're expected to test.

File: subsys/net/lib/websocket/websocket.c (hunk 1 above)
Comment (jukkar) [reply subsys/net/lib/websocket/websocket.c]: Ah, simple typo error. I will refactor this function or probably remove it altogether as it is now looking quite convoluted.

File: subsys/net/lib/websocket/websocket.c (hunk 2 above)
Comment (jukkar) [reply subsys/net/lib/websocket/websocket.c]: When unmasking data, we need to select correct byte from 32-bit mask value. This is an issue if we are reading less than 4 bytes. Anyway, I will refactor this this part in order to avoid these magic values.

File: subsys/net/lib/websocket/websocket.c (hunk 4 above)
Comment (jukkar) [reply subsys/net/lib/websocket/websocket.c]: I do not want to implement websocket server just to test this receive function. Using sockets would be nice of course but here sockets just complicate things unnecessarily. It should not really matter where the input data comes from to this function when testing it.

**Comment by jukkar** [comment]
New version that adds TX unit test and has a complete rewrite of the `websocket_recv_msg()` function. The recv function was too convoluted and not handling all the corner cases. While testing I also noticed that the sample app was doing things slightly wrong and it missed some error cases.

**Review by tbursztyka — APPROVED** [review]

**Comment by jukkar** [comment]
ping @nashif Any idea why sanity fails constantly in shippable? When running manually, I see no errors (tried multiple times):
```
$ sanitycheck -s tests/net/socket/websocket/net.socket.websocket -p native_posix_64 -c
Install the anytree module to use the --test-tree option
Install tabulate python module with pip to use --device-testing option.
Deleting output directory zephyr/sanity-out
INFO    - JOBS: 8
INFO    - Building initial testcase list...
INFO    - 1 test configurations selected, 730 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:    1/   1  100%  skipped:    0, failed:    0
INFO    - 1 of 1 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 5.19 seconds
INFO    - In total 742 test cases were executed on 1 out of total 220 platforms (0.45%)
```

**Comment by jukkar** [comment]
No functionality changes, just trying to tweak config options in order to get sanity to pass.
